src/backend/columnar/columnar_customscan.c

```diff
@@ -660,21 +816,84 @@ ColumnarScan_CreateCustomScanState(CustomScan *cscan)
 		sizeof(ColumnarScanState), T_CustomScanState);
 
 	CustomScanState *cscanstate = &columnarScanState->custom_scanstate;
-	cscanstate->methods = &ColumnarExecuteMethods;
+	cscanstate->methods = &ColumnarScanExecuteMethods;
 
-	if (EnableColumnarQualPushdown)
+	return (Node *) cscanstate;
+}
+
+
+static Node *
+EvalParamsMutator(Node *node, ExprContext *econtext)
+{
+	if (node == NULL)
 	{
-		columnarScanState->qual = cscan->scan.plan.qual;
+		return NULL;
 	}
 
-	return (Node *) cscanstate;
+	if (IsA(node, Param))
+	{
+		Param   *param = (Param *)node;
+		int16    typLen;
+		bool     typByVal;
+		Datum    pval;
+		bool     isnull;
+
+		get_typlenbyval(param->paramtype, &typLen, &typByVal);
+
+		/* XXX: should save ExprState for efficiency */
+		ExprState *exprState = ExecInitExprWithParams((Expr*) node, econtext->ecxt_param_list_info);
+		pval = ExecEvalExpr(exprState, econtext, &isnull);
+
+		return (Node *) makeConst(param->paramtype,
+								  param->paramtypmod,
+								  param->paramcollid,
+								  (int) typLen,
+								  pval,
+								  isnull,
+								  typByVal);
+	}
+
+	return expression_tree_mutator(node, EvalParamsMutator, (void *) econtext);
+}
+
+
+/*
+ * EvaluateClauseParams simply replaces Params with Consts in clauses using
+ * econtext.
+ */
+static List *
+EvaluateClauseParams(ExprContext *econtext, List *clauses)
+{
+	if (!EnableColumnarQualPushdown)
+	{
+		return NIL;
+	}
```

Comment:
although it functionally is the same, it feels a bit weird to check a GUC into this evaluate function. I think the code would be slightly easier to understand if the GUC check is on the callsites of this function. _not_ assigning anything to `columnarScanState->qual` if the GUC is disabled.

This would then also require the same check in `ColumnarScan_ReScanCustomScan` if my understanding is correct that we currently accidentally omit the result of `EvaluateClauseParams`

Reply:
Agreed; fixed.